pkg/gardenlet/controller/managedseed/managedseed_reconciler.go

```diff
@@ -0,0 +1,280 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package managedseed
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	gardencorev1beta1helper "github.com/gardener/gardener/pkg/apis/core/v1beta1/helper"
+	seedmanagementv1alpha1 "github.com/gardener/gardener/pkg/apis/seedmanagement/v1alpha1"
+	"github.com/gardener/gardener/pkg/client/kubernetes"
+	"github.com/gardener/gardener/pkg/client/kubernetes/clientmap"
+	"github.com/gardener/gardener/pkg/client/kubernetes/clientmap/keys"
+	"github.com/gardener/gardener/pkg/controllerutils"
+	"github.com/gardener/gardener/pkg/gardenlet/apis/config"
+	"github.com/gardener/gardener/pkg/logger"
+	"github.com/gardener/gardener/pkg/utils/imagevector"
+	kutil "github.com/gardener/gardener/pkg/utils/kubernetes"
+
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/retry"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// reconciler implements the reconcile.Reconcile interface for ManagedSeed reconciliation.
+type reconciler struct {
+	ctx         context.Context
+	clientMap   clientmap.ClientMap
+	config      *config.GardenletConfiguration
+	imageVector imagevector.ImageVector
+	recorder    record.EventRecorder
+	logger      *logrus.Logger
+}
+
+// newReconciler returns the new ManagedSeed reconciler.
+func newReconciler(ctx context.Context, clientMap clientmap.ClientMap, config *config.GardenletConfiguration, imageVector imagevector.ImageVector, recorder record.EventRecorder, logger *logrus.Logger) reconcile.Reconciler {
+	return &reconciler{
+		ctx:         ctx,
+		clientMap:   clientMap,
+		config:      config,
+		imageVector: imageVector,
+		recorder:    recorder,
+		logger:      logger,
+	}
+}
+
+func (r *reconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	gardenClient, err := r.clientMap.GetClient(r.ctx, keys.ForGarden())
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("could not get garden client: %w", err)
+	}
+
+	managedSeed := &seedmanagementv1alpha1.ManagedSeed{}
+	if err := gardenClient.Client().Get(r.ctx, request.NamespacedName, managedSeed); err != nil {
```

Comment:
If the `CachedRuntimeClients` feature gate is enabled, this call will be racy, because the underlying watch differs from the watch used for the event handlers (we had a similar problem in the BB/BE controller, see https://github.com/gardener/gardener/pull/2568 for more details).
TL;DR: we have to use the same watch/informer for the event handler and the get calls in the reconciler.

As switching to c-r informers for the event controllers is currently not an option, please use a lister in this reconciler for now. The [seed controller](https://github.com/gardener/gardener/blob/a2b77a84969703e42b0f18c3b305e73934b2b949/pkg/gardenlet/controller/seed/seed_control.go#L91) might be a good reference.

Reply:
Could I just use `DirectClient()`?

Reply:
Yeah, that's also an option. Probably even the better one, as it will be anyways refactored later on.

Reply:
Great, then this is what I'll do, thanks.